pkg/services/ssosettings/strategies/oauth_strategy.go

```diff
@@ -5,70 +5,102 @@ import (
 	"regexp"
 	"strings"
 
+	"github.com/grafana/grafana/pkg/login/social"
 	"github.com/grafana/grafana/pkg/services/ssosettings"
 	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util"
 )
 
 type OAuthStrategy struct {
-	provider                string
 	cfg                     *setting.Cfg
 	supportedProvidersRegex *regexp.Regexp
+	settingsByProvider      map[string]*social.OAuthInfo
+}
+
+var extraKeysByProvider = map[string][]string{
+	social.AzureADProviderName:      social.ExtraAzureADSettingKeys,
+	social.GenericOAuthProviderName: social.ExtraGenericOAuthSettingKeys,
+	social.GitHubProviderName:       social.ExtraGithubSettingKeys,
+	social.GrafanaComProviderName:   social.ExtraGrafanaComSettingKeys,
+	social.GrafanaNetProviderName:   social.ExtraGrafanaComSettingKeys,
 }
 
 var _ ssosettings.FallbackStrategy = (*OAuthStrategy)(nil)
 
 func NewOAuthStrategy(cfg *setting.Cfg) *OAuthStrategy {
 	compiledRegex := regexp.MustCompile(`^` + strings.Join(ssosettings.AllOAuthProviders, "|") + `$`)
-	return &OAuthStrategy{
+	oauthStrategy := &OAuthStrategy{
 		cfg:                     cfg,
 		supportedProvidersRegex: compiledRegex,
+		settingsByProvider:      make(map[string]*social.OAuthInfo),
 	}
+
+	oauthStrategy.loadAllSettings()
+	return oauthStrategy
 }
 
 func (s *OAuthStrategy) IsMatch(provider string) bool {
```

Comment:
Unrelated to this PR, but I think this can now be simplified:
```go
func (s *OAuthStrategy) IsMatch(provider string) bool {
  _, ok := s.settingsByProvider[provider]
  return ok
}